File: Problems 0100-0199/151. Reverse Words in a String/solution.cpp
class Solution {
    public:
        string reverseWords(string s) {
            reverse(s.begin(), s.end());
            int i = 0;
            int j = 0;
            int n = s.length();
            while (j < n) {
                while (j < n && s[j] == ' ') j++;
    
                if (j == n) break;
    
                if (i > 0) s[i++] = ' ';
    
                int l = j;
                int left = i;
                while (j < n && s[j] != ' ') {
                    s[i++] = s[j++];
                }
    
                int wordLen = j-l;
                reverse(s.begin() + left, s.begin() + left + wordLen);
            }
    
            s.resize(i);
            return s;
        }
    };

File: Problems 0100-0199/151. Reverse Words in a String/solution.py
class Solution:
    def reverseWords(self, s: str) -> str:
        chars = list(s)
        chars.reverse()

        n = len(chars)
        i = 0
        right = 0
        while i < n:
            while i < n and chars[i] == ' ':
                i += 1

            if i == n:
                break

            if right > 0:
                chars[right] = ' '
                right += 1
            
            left = right
            while i < n and chars[i] != ' ':
                chars[right] = chars[i]
                i += 1
                right += 1

            self.reverse(chars, left, right-1)
        
        chars = chars[:right]
        return "".join(chars)
        
    def reverse(self, s, l, r):
        while l < r:
            s[l], s[r] = s[r], s[l]
            l += 1
            r -= 1